### DISCUSSION TITLE
Improving indexing performance

### DISCUSSION THREAD

**Discussion by dinoschachten** [photoprism/photoprism#2183] [Q&A] [answered]
Hi everyone,

indexing of my existing directories has been running for about a day, at a speed of about 110 items per hour. Since my directories should contains some 100,00ish files, I'd very much like to improve that performance, especially given that none of the resources of my server and network seem to be anywhere close to their limits.

Here's what I know/did:

- My PhotoPrism server (Ubuntu, local Gigabit ethernet) has 24 virtual cores, all of which are running at below 25 %, except for some occasional spikes (around 50 %). The system has 16 GB of RAM, of which 30 % are used. The OS and Docker are running on an SSD, albeit an "older" one (Samsung EVO 840). Swap according to Ubuntu's System Monitor is at 14.2 % of 2 GiB.

- The files are stored on a NAS, the server mounts those SMB shares. Network usage on the server (receiving) is completely flatlining at 11.7 MB/s which seems a little odd, given that the NAS can deliver files to this server over SMB at up to 110+ MB/s, essentially maxing out the Gigabit ethernet.

- GPU load is between 10 and 30 % according to gpustat. I think I don't quite understand how the GPU is involved in the indexing by default. I believe I read it is not at all used unless dependencies according to the hardware are installed and set up? 

- According to the configuration options I can increase the number of workers - I have done that (hopefully without syntax errors), first to 12, then to 24 - and restarted of course. I can't really tell a difference.

- I did decrease the thumbnail quality and resolution which might have helped.

To summarise:
I'd like to utilise the resources of my server better to make indexing faster. My first approach/question would be to understand and configure the number of workers better as the documentation does not really explain them.
Any hints on other aspects are obviously most welcome, too - thanks in advance to all of you willing to share and help!

**Comment by dinoschachten** [comment, marked as answer]
Okay, here is the result of my further investigation - a bit embarrassing to be honest:

Apparently one of the network switches decided to randomly cap the bandwidth to 100 Mbit/s (the suspicion arose due to the continuous, never changing 11.7 MiB/s and was amplified by the fact that backups would suddenly run for ages, too).

After restarting both network switches along the path the performance is now splendid. 23 of my 24 cores running at 100 % continuously, RAM use fluctuates up to 14 GB, access to the NAS-hosted originals directories fluctuates between ~30 and ~110 MiB/s, which indicates that the read performance over network is good enough to leave the CPU as the bottleneck.

of course I should have tested right away if Gigabit speeds were available outside of Docker/PhotoPrism to find that the problem was completely unrelated to PhotoPrism.

That said, my temporary suspicion that this was related to Docker's lacking capabilities to utilise SMB mounts efficiently might hold some truth for Windows users, where this has been some cause for frustration: https://forums.docker.com/t/file-access-in-mounted-volumes-extremely-slow-cpu-bound/8076

I'll just leave this here in case others can in some way benefit from my "discussion" with myself.

**Reply by moeen-chowdhary** [reply]
Hello, @dinoschachten I would also like to know what is your average image file size and what performance level of indexing are you getting now after you have optimized your network throughput? Thanks.

**Comment by DJKeterson** [comment]
That was definitely slow, 110 items per hour is just crazy. I'm wondering what's the current speed now that you fixed the issue so I can compare.
Also which NAS are you using? And why are you mentioning the GPU load? I didn't think it would ever be used.

**Comment by dinoschachten** [comment]
Hi @moeen-chowdhary - sorry for the late response! @DJKeterson I'm so sorry, I completely missed that you wrote a year ago!!

I've got a couple of irregular files (like PSDs) that can take a while.

Previously indexed directories rush by within fractions of a second.

JPGs and HEICs from my phone: Around 2 per second.

On my current "server" (a 2017 laptop with very low power usage, which I prefer for 24/7 use), 50 MB uncompressed Sony ARW files take around 10 seconds each to convert, but since I am usually reviewing and selecting them in Lightroom on a local SSD, this is not really an issue. I believe it used to be significantly faster when the server's SSD wasn't as full as it is right now (I urgently need to switch to a larger drive for sidecar, thumbnails etc!).
